Fixed #18651 -- Enable optional assignments for simple tag.

prestontimmons opened the pull request.

timgraham commented on docs/howto/custom-template-tags.txt:
```diff
@@ -459,6 +459,18 @@ positional arguments. For example:
 
     {% my_tag 123 "abcd" book.title warning=message|lower profile=user.profile %}
 
+.. versionadded:: 1.8
```
1.9

timgraham commented on docs/howto/custom-template-tags.txt:
```diff
@@ -459,6 +459,18 @@ positional arguments. For example:
+It's possible to store the tag results in a template variable rather than
+directly outputting it. This is done by using the ``as`` argument followed by
+the variable name. Doing so enables you to output the content yourself where
+you see fit:
```
Isn't the idea more avoiding repeated calls to the same template tag if it's used multiple times?

prestontimmons replied:
Not strictly. The return value can be an iterable, or other type of object, in which case a variable is used later for a for loop or to access it's attributes.

timgraham commented on docs/howto/custom-template-tags.txt:
```diff
@@ -602,6 +614,11 @@ Assignment tags
 
 .. method:: django.template.Library.assignment_tag()
 
```
1.9

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -194,6 +196,15 @@ Miscellaneous
+assignment_tag()
+~~~~~~~~~~~~~~~~
+
+Django 1.4 added the :ref:`assignment_tag<howto-custom-template-tags-assignment-tags>`
```
prefer not linking to the deprecated thing as it's one less link to remove when thing is removed

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -194,6 +196,15 @@ Miscellaneous
+
+Django 1.4 added the :ref:`assignment_tag<howto-custom-template-tags-assignment-tags>`
+helper to ease the creation of template tags that store results in a template
+variable. The :ref:`simple_tag<howto-custom-template-tags-simple-tags>`
```
use :meth: link to simple_tag instead?

timgraham commented:
Missing tests for the new `simple_tag()` behavior?

prestontimmons commented:
Thanks for the feedback. I updated the docs and added tests. Let me know if you still think the wording should change about the ``as`` var.

As I looked through the ``assignment_tag`` tests, I noticed they were a duplicate of the ``simple_tag`` tests. Therefore, I removed all but a couple of those so the ``assignment_tag`` code path is still traversed.

I now wonder if this patch should maintain the current behavior of ``assignment_tag`` which raises a ``TemplateSyntaxError`` if ``as`` isn't specified? Personally, I've never understood the reason for that, but it should be easy to maintain if we want to.

Also, ``test_custom`` could use more cleanup. I'm inclined to do that, but don't want to mix that with this branch.

timgraham commented:
merged in cd4282816db9164791cd0ac97a3dc329ad92c522, thanks.
